research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -661,3 +721,24 @@ def shutdown(self) -> None:
         self.shutdown_dataloader(self.val_loader, "val_loader")
         self.shutdown_dataloader(self.test_loader, "test_loader")
         return super().shutdown()
+
+    def update_before_train(self, current_server_round: int) -> None:
+        # Was getting OOM errors that could only be fixed by manually cleaning up RAM
+        # https://github.com/pytorch/pytorch/issues/95462
+        # The above issue seems to be the situation I was in.
+        gc.collect()  # Cleans up unused variables
+        # As the linked issue above points out, calling gc.freeze() greatly reduces the
+        # overhead of garbage collection. (from 1.5s to 0.005s)
+        if current_server_round == 2:
+            # Collect runs even faster if we freeze after the end of the first iteration
+            # Likely because a lot of variables are created in the first pass
+            # If we freeze before the first pass, gc.collect has to check all those
+            # variables
+            gc.freeze()
```

Comment:
Maybe I have to look into the issue and details about the Python garbage collector further, but it looks like gc.freeze() is going to temporarily disable the automatic collection of unreachable objects by the garbage collector until gc.unfreeze() is called. If gc.freeze() is called, calls to gc.collect() only trigger the collection cycle is triggered but will not remove any objects. Are we calling gc.unfreeze() anywhere or is this somehow not required? Does gc.freeze() have anything to do with the OOM memory that is described or does it just make garbage collection faster? Why only on current_server_round=2 ?

Reply:
Maybe I should explain further in the comments, but gc.freeze doesn't stop gc.collect from working, it essentially takes a snapshot of what is currently in memory and makes it permanent. Any newly created objects after the freeze will still be cleaned up by gc.collect. The benefit of this is that gc.collect doesn't bother to check the objects that were frozen, therefore it only has to check items in memory that were created after the freeze, making things significantly faster. So the goal with gc.freeze is to ideally freeze only the objects that we are going to need in memory throughout the entire training process, so that gc.collect can focus on cleaning up anything that is only needed temporarily. Since we call collect right before the freeze, we can be pretty confident that most of whats in memory is going to be needed long term. The reason that we do this on current_server_round=2 is because we want to do the freeze after the first pass has occured. The first pass seems to create a lot of variables that we are going to need in memory throughout the entire training process, and so we would prefer those to be frozen so that gc doesn't have check them over and over every time we collect. The more items that gc is tracking, the slower the collect call is.

Reply:
gc unfreeze unfreezes the objects in memory that were frozen. So maybe it would be a good idea to unfreeze and then collect on train end in case the user has anything else they want to do. Not sure if that would clean up memory or not

Reply:
Ah I see so it will collect AND remove as per usual for object allocated after the gc.freeze() call. Seems like we would still want to call gc.unfreeze() and then gc.collect() at the end as you mentioned.

Reply:
If we just use gc.collect() without the gc.freeze() part does the OOM still happen? Or is this just for performance benefits?

Reply:
If we use only gc collect without freezing, it still resolves the OOM issue as the memory is still being cleaned up (particularly the generation 2 memory objects which are typically checked very infrequently). The calls to gc collect however take about 1.5s on my small dummy dataset. Using gc freeze reduces the overhead of gc collect to i think 0.005 seconds.

Reply:
I am little hesitant to add the additional complexity but if it's making a meaningful change on the actual dataset I am okay with it